tensorflow/lite/micro/kernels/batch_matmul.cc

```diff
@@ -0,0 +1,569 @@
+                       &op_data->quantization->output_shift);
+    // BatchMatMul has no fused activation functions. Therefore, set
+    // output activation min and max to min and max of int8_t or int16_t type.
+    if (lhs_data->type == kTfLiteInt8) {
+      op_data->quantization->output_activation_min =
+          std::numeric_limits<int8_t>::min();
+      op_data->quantization->output_activation_max =
+          std::numeric_limits<int8_t>::max();
+    } else {
+      op_data->quantization->output_activation_min =
+          std::numeric_limits<int16_t>::min();
+      op_data->quantization->output_activation_max =
+          std::numeric_limits<int16_t>::max();
+
+      TF_LITE_ENSURE_EQ(context, lhs_data->params.zero_point, 0);
+      TF_LITE_ENSURE_EQ(context, rhs_data->params.zero_point, 0);
+      TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+    }
+
+    op_data->quantization->lhs_zero_point = lhs_data->params.zero_point;
+    op_data->quantization->rhs_zero_point = rhs_data->params.zero_point;
+    op_data->quantization->output_zero_point = output->params.zero_point;
+  }
+
+  const int output_rank = std::max(lhs_rank, rhs_rank);
+  const RuntimeShape extended_lhs_shape =
+      RuntimeShape::ExtendedShape(output_rank, GetTensorShape(lhs_data));
+  const RuntimeShape extended_rhs_shape =
+      RuntimeShape::ExtendedShape(output_rank, GetTensorShape(rhs_data));
+
+  // Ensure any batch dimensions obey broacasting rules.
+  for (int i = 0; i < output_rank - 2; ++i) {
+    const int lhs_dim = extended_lhs_shape.Dims(i);
+    const int rhs_dim = extended_rhs_shape.Dims(i);
+    if (lhs_dim != rhs_dim) {
+      if (lhs_dim != 1) {
+        TF_LITE_ENSURE_EQ(context, rhs_dim, 1);
+      }
+    }
+  }
+  bool adj_x = op_context.params->adj_x;
+  bool adj_y = op_context.params->adj_y;
+  // Ensure other dimensions work for matrix multiplication.
+  int accum_dim_lhs = adj_x ? extended_lhs_shape.Dims(output_rank - 2)
+                            : extended_lhs_shape.Dims(output_rank - 1);
+  int accum_dim_rhs = adj_y ? extended_rhs_shape.Dims(output_rank - 1)
+                            : extended_rhs_shape.Dims(output_rank - 2);
+
+  TF_LITE_ENSURE_EQ(context, accum_dim_lhs, accum_dim_rhs);
+  TfLiteStatus status =
+      ReshapeOutputTensor(context, node, extended_lhs_shape, extended_rhs_shape,
+                          adj_x, adj_y, output_rank, output);
+  return status;
+}
+
+void* Init(TfLiteContext* context, const char* buffer, size_t length) {
+  // This is a builtin op, so we don't use the contents in 'buffer', if any.
+  // Instead, we allocate a new object to carry information from Prepare() to
+  // Eval().
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+  MicroContext* micro_context = GetMicroContext(context);
+  return micro_context->AllocatePersistentBuffer(sizeof(OpData));
+}
+
+TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
+  return CalculateOpData(context, node);
```

Comment:
This is the requested way for porting kernel operators (from 2021 onwards).

Reply:
Having a function can be useful when coffee can be shared between reference and optimized kernels. Since this PR is only adding the reference kernel, following @rascani's suggestion of putting the cube directly in prepare is a good idea.

Refactoring can be done (if needed) when the optimized kernels are implemented.

Reply:
updated